src/txmempool.cpp

```diff
@@ -496,7 +497,7 @@ void CTxMemPool::addUnchecked(const CTxMemPoolEntry &entry, setEntries &setAnces
     CAmount delta{0};
     ApplyDelta(entry.GetTx().GetHash(), delta);
     if (delta) {
-            mapTx.modify(newit, update_fee_delta(delta));
+        mapTx.modify(newit, update_modify_fee(delta));
```

Comment:
This looks wrong.

Lets say `nFee` is `100`.

Before this PR we would store `feeDelta=5` to designate a modified fee of `105`. Lets say this code wants to change that to a modified fee of `107`. Then this `delta` argument here would be `7`. In this case, inside the old `CTxMemPoolEntry::UpdateFeeDelta()` the variables `nModFeesWithDescendants` and `nModFeesWithAncestors` would be incremented with `2`, and `7` would be stored in `feeDelta` (to designate a modified fee of `107` (`nFee=100` plus `feeDelta=7`).

After this PR we would store `modified_fee=105` to designate a modified fee of `105`. Then this code gets executed with argument `delta=7`. The new `CTxMemPoolEntry::UpdateModifyFee()` would increment the variables `nModFeesWithDescendants` and `nModFeesWithAncestors` with `7` (not with `2` as before) and would store `105+7=112` in `modified_fee` to designate a modified fee of `112` (not `107`).

Either this call site, or the method `UpdateModifyFee()` has to be changed in order to keep the same behavior, or I miss something. I think the other call site of `UpdateModifyFee()` is correct.

Reply:
This code path is for newly added txs. Thus, `nFee` is always equal to `modified_fee` before this call to update the fee. It is not possible to execute this code for txs that have already been added, whose fee is modified repeatedly.

Luckily we have tests to catch this.

If I am missing something, it might be good if you write a test (or explain steps to reproduce).

Reply:
Ok, so the behavior of the code before and after this PR is the same only in the case where `feeDelta` is `0` (old code) / `nFee` is equal to `modified_fee` (new code). It is quite obscure to me whether this code is executed only for such cases (and will always be in the future). What about adding an assert to ensure that:

```diff
     if (delta) {
-        mapTx.modify(newit, [&delta](CTxMemPoolEntry& e) { e.UpdateModifiedFee(delta); });
+        mapTx.modify(newit, [&delta](CTxMemPoolEntry& e) {
+            assert(e.GetFee() == e.GetModifiedFee());
+            e.UpdateModifiedFee(delta);
+        });
     }
```

or, to preserve the behavior in all cases, not just when `feeDelta=0`:

```diff
     if (delta) {
-        mapTx.modify(newit, [&delta](CTxMemPoolEntry& e) { e.UpdateModifiedFee(delta); });
+        mapTx.modify(newit, [&delta](CTxMemPoolEntry& e) {
+            e.UpdateModifiedFee(e.GetModifiedFee() - e.GetFee() + delta);
+        });
     }
```

I would be more comfortable with the second variant as it preserves the behavior in all cases and there is no chance that a newly added assert is too strong and is triggered unintentionally by a legit code.

Reply:
I don't like the second variant, because it makes the code more verbose and confusing. Also, it will lead to signed integer overflow (the thing that this pull is trying to solve).

Reply:
I've pushed a documention update to make it clearer along with an `Assume`. I don't think an `assert` is the right thing to pick here. It seems incredibly unlikely that someone is going to modify the code while violating the assumption and at the same time have passing tests (hopefully added in the same patch that modifies the code). So I am also happy to remove the `Assume` again, if deemed useless.

Reply:
Ok, IMO it is better with the `Assume()` compared to without any protection.

This call site is kind of misusing the `UpdateModifiedFee()` method whose semantic is to alter the existent value with a provided delta, while this call site is actually trying to _set_ the value. It works as long as the value is `0` here, before the call.

What about introducing a new method to set the modified fee and use it here?

```cpp
void CTxMemPoolEntry::SetModifiedFee(CAmount modified_fee_)
{
    nModFeesWithDescendants = modified_fee_;
    nModFeesWithAncestors = modified_fee_;
    modified_fee = modified_fee_;
}
```

Or check that what we intended was actually done correctly afterwards:

```diff
     if (delta) {
         mapTx.modify(newit, [&delta](CTxMemPoolEntry& e) { e.UpdateModifiedFee(delta); });
+        Assume(entry.GetModifiedFee() == SaturatingAdd(entry.GetFee(), delta));
     }
```

Reply:
I think it is fine to call `UpdateModifiedFee` to set the initial fee update. Introducing a new method doesn't allow to get rid of the Assumption that the fee wasn't previously modified.

Happy to move the `Assume`/comment after the mapTx.modify call.

Reply:
Closing thread for now. Let me know if I should reopen.